packages/core/src/lib/base_protocol.ts

```diff
@@ -28,13 +29,26 @@ export class BaseProtocol {
   }
 
   protected async getPeer(peerId?: PeerId): Promise<Peer> {
-    const { peer } = await selectPeerForProtocol(
-      this.peerStore,
-      [this.multicodec],
-      peerId
-    );
+    let peer = await this._getPeer(peerId);
+    if (!peer) {
+      await waitForRemotePeer(this.multicodec, this.peerStore);
+      peer = (await this._getPeer(peerId)) as Peer;
```

Comment:
This is duplicate work. `waitForRemotePeer` as already found a peer, it should be the one we use.